### PR TITLE
[TRTLLM-8831][feat] Enable early exit with overlap scheduler

### PR DESCRIPTION
**Pull request by Funatiq**
## Summary by CodeRabbit

* **New Features**
  * Exposed `willCompleteNextIteration` method for improved request state querying.
  * Enhanced streaming state transitions with support for intermediate completion states.
  * Added per-request control over last-token exclusion in generation logits.

* **Bug Fixes**
  * Fixed overlap scheduler inadvertently increasing token counts in certain execution paths.

* **Tests**
  * Improved per-sequence output validation and logits verification in test suites.

## Description

Usually with overlap scheduler we schedule one iteration ahead. The idea here is that when a request reaches its `max_tokens` we can avoid the last extra iteration because we know that the request will finish. This also leads to a final iteration of the execution loop where no requests are scheduled (`scheduled_batch` is empty) but we still have to process the results from the previous iteration (`previous_batch` exists).

- Update `PyExecutor` to set `GENERATION_TO_COMPLETE` state for requests that will complete next iteration.
- Update `MicroBatchScheduler` bindings to skip scheduling after `GENERATION_TO_COMPLETE` state.
- Fix `_executor_loop_overlap` to finish previous batch if current batch is empty.
- Move iter_counter handling from `ModelEngine` to `PyExecutor` to account for empty batches.
- Fix logits output length with early exit.
- Fix call counts with early exit in tests.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by lucaslie — APPROVED** [review]

File: tests/unittest/llmapi/test_llm_pytorch.py [hunk 1]
```diff
@@ -182,28 +182,30 @@ def test_llm_reward_model():
...
-    assert kv_cache_metrics.num_total_allocated_blocks == 1
-    assert kv_cache_metrics.num_new_allocated_blocks == 1
-    assert kv_cache_metrics.num_reused_blocks == 0
-    assert kv_cache_metrics.num_missed_blocks == 1
-    assert kv_cache_metrics.kv_cache_hit_rate == 0
-
-    assert perf_metrics.first_iter is not None
-    assert perf_metrics.iter - perf_metrics.first_iter == sampling_params.max_tokens - 1
-    assert perf_metrics.last_iter == perf_metrics.iter
+    with LLM(model=llama_model_path,
```
Comment (achartier) [inline tests/unittest/llmapi/test_llm_pytorch.py]: Looks like unrelated refactoring?

**Review by achartier — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 2]
```diff
@@ -1394,17 +1400,18 @@ def _executor_loop_overlap(self):
...
+                previous_tensors = self.previous_batch and self.previous_batch.sample_state
+                # If there are previous draft tokens, we need to update the target requests to accept some draft tokens.
+                # When there's any accepted tokens, we can't directly use the previous batch's outputs in this iteration for the target model,
+                # so we'll set the target model's input to None and skip updating the target requests after target model forward.
+                use_previous_draft_tokens = self.has_previous_draft_tokens
+                if self.drafter is not None and (self.use_spec_decode or
+                                                    use_previous_draft_tokens):
+                    target_inputs = self._handle_speculative_decoding(
+                        scheduled_batch, previous_tensors,
+                        previous_tensors_device)
```
[the remark this answers is not in the document]
Comment (Funatiq) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: Thanks for confirming, I added the change.
